File: delta/latest.ts
import { Composer, Context, InlineKeyboard } from "../deps.ts";
import type { Release } from "../types/Github.d.ts";
import { last } from "../utils/generator.ts";
import hecker from "../utils/checker.ts";

const composer = new Composer();

export const message = async (data: Release) =>
  // make a message about the release date of the new release
  `Latest version is <b><a href="${
    (await hecker(data.tag_name, data.body)).url
  }">${data.tag_name}</a></b> released at <code>${
    new Date(data.published_at).toDateString()
  }</code> by <a href="${data.author.html_url}">${data.author.login}</a>`;

export const keyboard = (data: Release) =>
  new InlineKeyboard().url("More Information", data.html_url);

composer.command("last", async (ctx: Context): Promise<void> => {
  const req = await last();
  await ctx.reply(await message(req), {
    parse_mode: "HTML",
    reply_markup: keyboard(req),
  });
});

export default composer;

File: src/functions/latest.rs
use crate::utils::{github::GitHub, keyboard::Keyboard, message::Rustina};
use octocrab::models::repos::Release;
use teloxide::{
    payloads::SendMessageSetters,
    prelude::*,
    types::{InlineKeyboardMarkup, ParseMode},
};

pub async fn command(bot: &Bot, github: GitHub, msg: &Message) -> ResponseResult<()> {
    let latest = github.get_latest().await.unwrap();

    bot.send_message_tf(msg.chat.id, view(&latest), msg)
        .parse_mode(ParseMode::Html)
        .reply_markup(keyboard(&latest))
        .await?;

    Ok(())
}

pub fn view(release: &Release) -> String {
    format!(
        "<b>Latest version is <a href=\"https://releases.rs/docs/{}\">\
        {}</a> released at </b> <code>{}</code> <b>by</b> <a href=\"{}\">\
        {}</a>.\
        \n\n\
        <b>In order to install latest update, execute</b> <code>rustup update</code> <b>on your terminal!</b>
        ",
        release.tag_name,
        release.tag_name,
        release.published_at.unwrap().date_naive(),
        release.author.html_url,
        release.author.login,
    )
}

pub fn keyboard(release: &Release) -> InlineKeyboardMarkup {
    let mut keyboard = Keyboard::new();
    keyboard.url("More Information", release.html_url.as_str())
}
